shared/src/main/kotlin/com/egm/stellio/shared/util/GeoQueryUtils.kt

```diff
@@ -1,61 +1,143 @@
+const val GEO_QUERY_PARAM_GEOPROPERTY = "geoproperty"
+
+const val GEO_QUERY_GEOREL_NEAR = "near"
+const val GEO_QUERY_GEOREL_WITHIN = "within"
+const val GEO_QUERY_GEOREL_CONTAINS = "contains"
+const val GEO_QUERY_GEOREL_INTERSECTS = "intersects"
+const val GEO_QUERY_GEOREL_EQUALS = "equals"
+const val GEO_QUERY_GEOREL_DISJOINT = "disjoint"
+const val GEO_QUERY_GEOREL_OVERLAPS = "overlaps"
+val GEO_QUERY_ALL_GEORELS = listOf(
+    GEO_QUERY_GEOREL_NEAR,
+    GEO_QUERY_GEOREL_WITHIN,
+    GEO_QUERY_GEOREL_CONTAINS,
+    GEO_QUERY_GEOREL_INTERSECTS,
+    GEO_QUERY_GEOREL_EQUALS,
+    GEO_QUERY_GEOREL_DISJOINT,
+    GEO_QUERY_GEOREL_OVERLAPS
+)
+const val GEOREL_NEAR_DISTANCE_MODIFIER = "distance"
+const val GEOREL_NEAR_MAXDISTANCE_MODIFIER = "maxDistance"
+const val GEOREL_NEAR_MINDISTANCE_MODIFIER = "minDistance"
+
+private val georelNearRegex = "^near;(?:minDistance|maxDistance)==\\d+$".toRegex()
 
-fun parseGeoQueryParameters(
-    requestParams: MultiValueMap<String, String>,
+suspend fun parseGeoQueryParameters(
+    requestParams: Map<String, String>,
     contextLink: String
-): Either<APIException, GeoQuery> {
-    val georel = requestParams.getFirst(GeoQueryUtils.GEO_QUERY_PARAM_GEOREL)
-    val geometry = requestParams.getFirst(GeoQueryUtils.GEO_QUERY_PARAM_GEOMETRY)
-    val coordinates = requestParams.getFirst(GeoQueryUtils.GEO_QUERY_PARAM_COORDINATES)?.decode()
-    val geoproperty = requestParams.getFirst(GeoQueryUtils.GEO_QUERY_PARAM_GEOPROPERTY)?.let {
-        JsonLdUtils.expandJsonLdTerm(it, contextLink)
+): Either<APIException, GeoQuery?> = parseGeoQueryParameters(requestParams, listOf(contextLink))
+
+suspend fun parseGeoQueryParameters(
+    requestParams: Map<String, String>,
+    contexts: List<String>
+): Either<APIException, GeoQuery?> = either {
+    val georel = requestParams[GEO_QUERY_PARAM_GEOREL]?.also {
+        checkGeorelParam(it).bind()
+    }
+    val geometry = requestParams[GEO_QUERY_PARAM_GEOMETRY]?.let {
+        if (GeoQuery.GeometryType.isSupportedType(it))
+            GeoQuery.GeometryType.forType(it).right()
+        else
+            BadRequestDataException("$it is not a recognized value for 'geometry' parameter").left()
+    }?.bind()
+    val coordinates = requestParams[GEO_QUERY_PARAM_COORDINATES]?.decode()?.let {
+        stringifyCoordinates(it)
+    }
+    val geoproperty = requestParams[GEO_QUERY_PARAM_GEOPROPERTY]?.let {
+        expandJsonLdTerm(it, contexts)
     } ?: JsonLdUtils.NGSILD_LOCATION_PROPERTY
 
-    return GeoQuery(
-        georel = georel,
-        geometry = geometry,
-        coordinates = coordinates,
-        geoproperty = geoproperty
-    ).right()
+    // if at least one parameter is provided, the three must be provided for the geoquery to be valid
+    val notNullGeoParameters = listOfNotNull(georel, geometry, coordinates)
+    if (notNullGeoParameters.isEmpty())
+        null
+    else if (georel == null || geometry == null || coordinates == null)
```

Comment:
This check is the same than above ?

Reply:
- 1st check: if empty, no geo parameter has been provided, geoquery is null
- 2nd check: not empty so at least one geo parameter has been provided. in this case, `georel`, `geometry` and `coordinates` must be provided